core/types/src/attestation_key.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Attestation Key types
+
+use crate::{
+    attestation_key::QuoteSignature::{Linkable, UnLinkable},
+    new_type_accessors_impls,
+    report::{ExtendedProductId, FamilyId},
+    svn::ConfigId,
+    FfiError,
+};
+use mc_sgx_core_sys_types::{sgx_att_key_id_ext_t, sgx_ql_att_key_id_t, sgx_quote_sign_type_t};
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum MrSignerKeyHash {
+    Sha256([u8; 32]),
+    Sha384([u8; 48]),
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct AlgorithmId(u32);
+
+new_type_accessors_impls! {
+    AlgorithmId, u32;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Version(u16);
+
+new_type_accessors_impls! {
+    Version, u16;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Id(u16);
+
+new_type_accessors_impls! {
+    Id, u16;
+}
+
+/// The type of quote signature.  Only valid for EPID quotes.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+#[non_exhaustive]
+#[repr(u16)]
+enum QuoteSignature {
+    UnLinkable,
+    Linkable,
+}
+
+impl From<sgx_quote_sign_type_t> for QuoteSignature {
+    fn from(sign_type: sgx_quote_sign_type_t) -> QuoteSignature {
+        match sign_type {
+            // Per the header `sgx_quote.h` and
+            // https://download.01.org/intel-sgx/sgx-linux/2.17.1/docs/Intel_SGX_Developer_Reference_Linux_2.17.1_Open_Source.pdf
+            // the `sgx_att_key_id_ext_t::att_key_type` is only valid for EPID
+            // quotes it will be 0 otherwise, which also happens to map to
+            // SGX_UNLINKABLE_SIGNATURE.
+            sgx_quote_sign_type_t::SGX_UNLINKABLE_SIGNATURE => UnLinkable,
+            sgx_quote_sign_type_t::SGX_LINKABLE_SIGNATURE => Linkable,
+            v => panic!("Got unknown 'sgx_quote_sign_type_t': {:?}", v),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct QuoteLibAttestationKeyId(sgx_ql_att_key_id_t);
+
+impl QuoteLibAttestationKeyId {
+    /// The ID
+    pub fn id(&self) -> Id {
+        self.0.id.into()
+    }
+
+    /// The key version
+    pub fn version(&self) -> Version {
+        self.0.version.into()
+    }
+
+    /// The hash of the MRSIGNER key.
+    ///
+    /// # Errors
+    /// When the length of the hash is not a valid length for a hash,
+    /// [`sgx_ql_att_key_id_t::mrsigner_length`].
+    pub fn mr_signer_key_hash(&self) -> Result<MrSignerKeyHash, FfiError> {
```

Comment:
Not fond of the asymmetry with this function being fallible, but it seems at some point we have to be able to fail converting from the _loose_ C definition to a more strict rust one.

Reply:
Is this something we could catch at the time of constructing a `QuoteLibAttestationKeyId`, instead of later on when we try to access the `SignerKeyHash`?

Reply:
My initial thought:
- Re-work `new_type_accessors_impls!` to be `TryFrom` instead of `From`, or perhaps 2 versions for things that can't fail
- For the possible failure types add a trait that has a `validate()` method which returns `Result<Self, FfiError>`
I'm thinking this should probably be a follow on issue, because I bet it will touch a bit of code

Reply:
My thought here was that instead of having `struct QuoteLibAttestationKeyId(sgx_quote_lib_attestation_key_id_t)`, we could have something like `struct QuoteLibAttestationKeyId { inner: sgx_quote_lib_attestation_key_id_t }` and then `impl TryFrom<sgx_quote_lib_attestation_key_id_t> for QuoteLibAttestationKeyId { /* ... */ }`. That way we can never construct an invalid `QuoteLibAttestationKeyId`. The tuple struct newtype implicitly allows us to construct a `QuoteLibAttestationKeyId` from _any_ `sgx_quote_lib_attestation_key_id_t`, which is not the semantics we actually want, IMO.

I somewhat agree that this might want to be a follow-on issue, as I suspect this pattern of being able to construct invalid versions of the Rust wrapper types might exist in other places as well.

Reply:
Actually on looking into it further it's not the tuple type that causes this but (as you correctly mentioned) the `new_type_accessors_impls` macro, which I don't really like in general as it gives users of the Rust types a tremendous amount of access to the FFI types. We're already using it a lot though and it is very convenient, so... not sure what the right call here is. It would _definitely_ be a follow on issue to make any changes to how we're using that macro.

Reply:
>  it gives users of the Rust types a tremendous amount of access to the FFI types
Looking at it, it seems that the `AsMut` is the main thing that seems out of place, if we want to ensure that the rust type is always valid.  I think the rest can stay.
We'll want the `AsRef` to make it easier to pass to the C interfaces:
```rust
let mut rust_version = MyRustType::default();
rust_version.some_setter_method(value);
...
let result = unsafe{ a_pass_by_pointer_c_function(rust_version.as_ref()) }
```
pretty sure the syntax isn't 100%, but hopefully it communicates the idea